File: modules/rest/src/main/java/dev/httpmarco/polocloud/modules/rest/socket/web/WebSocketService.java
package dev.httpmarco.polocloud.modules.rest.socket.web;

import dev.httpmarco.polocloud.modules.rest.RestModule;
import dev.httpmarco.polocloud.modules.rest.socket.WebSocket;
import dev.httpmarco.polocloud.modules.rest.socket.web.impl.v1.ConsoleLogWebWebSocket;
import dev.httpmarco.polocloud.modules.rest.socket.web.impl.v1.NotifyWebSocket;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.util.ArrayList;
import java.util.List;

@Accessors(fluent = true)
public class WebSocketService {

    public static final String API_PATH = "/polocloud/api/v1";
    private final RestModule restModule;
    private final WebSocketAuthService webSocketAuthService;
    @Getter
    private final List<WebSocket> webSockets;

    public WebSocketService(RestModule restModule) {
        this.restModule = restModule;
        this.webSocketAuthService = new WebSocketAuthService(this.restModule);
        this.webSockets = new ArrayList<>();

        invoke();
    }

    private void registerWebSockets(WebSocket... webSocket) {
        this.webSockets.addAll(List.of(webSocket));
    }

    public void invoke() {
        registerWebSockets(
                new NotifyWebSocket(this.restModule),
                new ConsoleLogWebWebSocket(this.restModule)
        );

        for (var webSocket : this.webSockets) {
            this.restModule.app().ws(API_PATH + webSocket.path(), ws -> {
                ws.onConnect(context -> this.webSocketAuthService.authenticate(webSocket, context));
                ws.onClose(webSocket::onClose);
                ws.onMessage(webSocket::onMessage);
                ws.onError(webSocket::onError);
            });
        }
    }
}

File: modules/rest/src/main/kotlin/dev/httpmarco/polocloud/modules/rest/socket/web/WebSocketService.kt
package dev.httpmarco.polocloud.modules.rest.socket.web

import dev.httpmarco.polocloud.modules.rest.RestModule
import dev.httpmarco.polocloud.modules.rest.socket.BaseWebSocket
import dev.httpmarco.polocloud.modules.rest.socket.web.impl.v3.AliveWebSocket
import dev.httpmarco.polocloud.modules.rest.socket.web.impl.v3.service.EventWebSocket
import dev.httpmarco.polocloud.modules.rest.socket.web.impl.v3.PolocloudLogWebSocket
import dev.httpmarco.polocloud.modules.rest.socket.web.impl.v3.service.ServiceScreenWebSocket

class WebSocketService {

    companion object {
        const val API_PATH = "/polocloud/api/v3"
    }

    private val authService = WebSocketAuthService()
    private val sockets = mutableListOf<BaseWebSocket>()

    init {
        register(
            AliveWebSocket(),
            PolocloudLogWebSocket(),
            ServiceScreenWebSocket(),
            EventWebSocket()
        )

        this.sockets.forEach { socket ->
            RestModule.instance.httpServer.app.ws(API_PATH + socket.path) { ws ->
                ws.onConnect { context -> authService.authenticate(socket, context) }
                ws.onClose { socket::onClose }
                ws.onMessage { socket::onMessage }
                ws.onError { socket::onError }
            }
        }
    }

    private fun register(vararg webSockets: BaseWebSocket) {
        sockets += webSockets
    }

    fun shutdown() {
        this.sockets.forEach { socket ->
            socket.shutdown()
        }
    }
}